### ISSUE TITLE
DateTime conversion of parameter in format 'yyyy-MM-dd'

### ISSUE THREAD

**Issue by ChiefWiggum** [aspnet/Routing#237]
Hi guys,

we are having an issue with urls containing a date in the format `yyyy-MM-dd`. I get an entry in the `RouteData.Values` with a type of `DateTime`, so far so good, but if the date is '2015-11-23' it gets converted into a `DateTime(yyyy,MM,dd)` (the expected result) while the date '2015-11-03' gets converted into a `DateTime(yyyy,dd,MM)` (he horrible US date format).
Currently my action has just one parameter `DateTime date`. I thought when I change the type from `DateTime` to `string` I would get a nice little string in the format `yyyy-MM-dd`, but no, it takes the `DateTime` value in the `RouteData` and makes a `ToString()` which then results culture specific string.
In short:
- localhost/2015-11-23 => date = new DateTime(2015,11,23)
- localhost/2015-11-03 => date = new DateTime(2015,3,11)

Has anyone an idea what I need to change to either
- Get my date parameter as a string in the given format so I can parse it myself or
- Pass in some formatting somewhere so it always converts the urls with my given format.

Cheers

**Comment by ChiefWiggum** [comment]
I found one issue, but it's still not fully resolved. I hadn't realized that one of my collegues has implemented his own `IDynamicRouter` which included his own `DateTime` checking & parsing. The weird thing is that now when I debug my Action the `DateTime date` parameter has the wrong date while the RoutData.Values contains a date with the correct data. How can my RouteData.Values be correct and the parameter value of the action be wrong at the same time?

**Comment by ChiefWiggum** [comment]
I assume we're doing something wrong when invoking the action, any ideas?

        public async Task RouteAsync(RouteContext context)
        {
            var services = context.HttpContext.RequestServices;
            var segments = Parser.ParseSegments(context);

            var match = Tree.Search(segments);
            if (!match.Success)
                return;

            MapRouteData(context, match);

            var actionselector = services.GetRequiredService<IActionSelector>();
            var accessor = services.GetRequiredService<IActionContextAccessor>();

            var action = actionselector.SelectAsync(context);
            accessor.ActionContext = new ActionContext(context.HttpContext, context.RouteData, action.Result);

            var actionInvokerFactory = context.HttpContext.RequestServices.GetRequiredService<IActionInvokerFactory>();
            var invoker = actionInvokerFactory.CreateInvoker(accessor.ActionContext);

            await invoker.InvokeAsync();

            context.IsHandled = true;
        }

        private void MapRouteData(RouteContext context, DecisionTreeSearchMatch match)
        {
            foreach (var node in match.UrlNodes)
            {
                foreach (var keyValuePair in node.GetRouteData(match.UrlNodes))
                {
                    context.RouteData.Values.Add(keyValuePair);
                }
            }

            foreach (var values in context.HttpContext.Request.Query)
            {
                if(!context.RouteData.Values.ContainsKey(values.Key))
                    context.RouteData.Values.Add(values.Key, values.Value);
            }
        }

**Comment by rynowak** [comment]
RouteData values in MVC are parsed according to the `InvariantCulture` rule, which (as I remember) are similar to the US English culture.

Try an experiment - set up an HTML form that passes a `DateTime` value with your desired format - does it work as you expect? (Form uses `CurrentCulture`). Query also parses values according to `CurrentCulture`, but I noticed in your sample that you are copying Query values into RouteValues. RouteValues has the highest precedence.

If you want to use a different culture for parsing route values, replace `RouteValueProviderFactory` with your own.
